runtime.md

```diff
@@ -67,38 +62,55 @@ Note: these operations are not specifying any command-line APIs, and the paramen
+
+Any error generated during the creation of the container MUST generate an error and a new container MUST not be created.
+
+Runtime callers who are interested in pre-create validation can run [bundle-validation tools](implementations.md#testing--tools) before invoking the start operation.
+
+### Start
+`start <container-id>`
 
-The runtime MAY validate `config.json` against this spec, either generically or with respect to the local system capabilities, before creating the container ([step 2](#lifecycle)).
-If the runtime does not perform initial validation and triggers an error due to an invalid or incompatible configuration, it MUST generate an error and jump to cleanup ([step 7](#lifecycle)).
-Runtime callers who are interested in pre-start validation can run [bundle-validation tools](implementations.md#testing--tools) before invoking the start operation.
+This operation MUST generate an error if it is not provided the container ID.
+Attempting to start a container that does not exist MUST generate an error.
+Attempting to start an already started container MUST have no effect on the container and MUST generate an error.
+This operation MUST run the user-specified code as specified by [`process`](config.md#process-configuration).
+If the runtime fails to run code as specified, an error MUST be generated.
+This process MUST be run in the PID namespace associated with the container.
 
-Attempting to start an already running container MUST have no effect on the container and MUST generate an error.
+### Kill
+`kill <container-id> <signal>`
 
-### Stop
+This operation MUST generate an error if it is not provided the container ID.
+Attempting to send a signal to a container that is not running MUST have no effect on the container and MUST generate an error.
+This operation MUST send the specified signal to the process in the container.
 
-`stop <container-id>`
+### Delete
+`delete <container-id>`
 
 This operation MUST generate an error if it is not provided the container ID.
-This operation MUST stop and delete a running container.
-Stopping a container MUST stop all of the processes running within the scope of the container.
-Deleting a container MUST delete the associated namespaces and resources associated with the container.
-Once a container is deleted, its `id` MAY be used by subsequent containers.
-Attempting to stop a container that is not running MUST have no effect on the container and MUST generate an error.
+Attempting to delete a container that does not exist MUST generate an error.
+Attempting to delete a container whose process is still running MUST first stop the process by invoking the `kill` operation on the container.
+If, after the `kill` operation is invoked, the process does not stop then an error MUST be generated and the container MUST NOT be deleted.
```

Comment:
On Thu, May 26, 2016 at 09:54:12AM -0700, Daniel, Dao Quang Minh wrote:
> > +If, after the `kill` operation is invoked, the process does not
> > stop then an error MUST be generated and the container MUST NOT be
> > deleted.
>
> This seems a bit weird. Usually we uses two signals here, SIGTERM
> first ( or user-provided signal ) with a timeout, and finally
> SIGKILL. I think we can change this term to something similar to "if
> the runtime cant stop a container then an error MUST be generated
> and the container MUST NOT be deleted."

I'd rather stay away from implicit “please exit cleanly” behavior
(e.g. sending SIGTERM without having it mentioned in the spec).  If we
are sending SIGTERMS, we need to tell config authors that explicitly
so they can prepare their container to catch and handle that signal
appropriately [1].  On the other hand, Windows doesn't seem to support
SIGTERM/SIGKILL [2].  So I think we want a:

  terminate [--force]

operation (or ‘stop [--force]’, or ‘kill [--soft]’, or whatever) that
callers can use independent of platform.  On Linux we'd require the
soft version to just send SIGTERM, and on Windows we'd require it to
do whatever hcsshim.ShutdownComputeSystem does [2].  There's no need
to specify what we do for hard kills, because the container code has
no opportunity to handle those.

[1]: https://github.com/opencontainers/runtime-spec/issues/356#issuecomment-209582923
[2]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/PlGKu7QUwLE
     Subject: Fwd: Windows support for OCI stop/signal/kill (runtime-spec#356)
     Date: Thu, 26 May 2016 11:03:29 -0700
     Message-ID: <20160526180329.GL17496@odin.tremily.us>